csharp/autobuilder/Semmle.Autobuild.CSharp/DotNetRule.cs

```diff
@@ -147,11 +140,59 @@ private static BuildScript DownloadDotNet(IAutobuilder<AutobuildOptionsShared> b
                     continue;
                 }
 
-                installScript &= DownloadDotNetVersion(builder, installDir, version);
-                validGlobalJson = true;
+                return version;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Returns a script for downloading relevant versions of the
+        /// .NET Core SDK. The SDK(s) will be installed at <code>installDir</code>
+        /// (provided that the script succeeds).
+        /// </summary>
+        private static BuildScript DownloadDotNet(IAutobuilder<AutobuildOptionsShared> builder, string installDir)
+        {
+            if (!string.IsNullOrEmpty(builder.Options.DotNetVersion))
+                // Specific version supplied in configuration: always use that
+                return DownloadDotNetVersion(builder, installDir, builder.Options.DotNetVersion);
+
+            // Download versions mentioned in `global.json` files
+            // See https://docs.microsoft.com/en-us/dotnet/core/tools/global-json
+            var installScript = BuildScript.Success;
+            var foundFrameworkVersion = false;
+            var globalJsonSdkVersion = SdkVersionFromGlobalJson(builder);
+
+            if (!string.IsNullOrEmpty(globalJsonSdkVersion))
+            {
+                installScript &= DownloadDotNetVersion(builder, installDir, globalJsonSdkVersion);
+                foundFrameworkVersion = true;
+            }
+            else
+            {
+                // If there is no `global.json` to retrieve the .NET version from, we find all the
+                // TargetFramework values in the project files and install the latest versions
+                // from the corresponding channels
```

Comment:
Why do we only install the latest? Why not all? If there are multiple projects with different target framework settings in the repo, some of them will fail the build, won't it?

Reply:
We do install all! Maybe you had a look at the initial draft version of this PR, which only installed the latest, but now the review-ready version does install all the ones we find (in theory at least - pending changes we may have to make to `DownloadDotNetVersion`)

Reply:
> If there are multiple projects with different target framework settings in the repo, some of them will fail the build, won't it?

This is an interesting question. I would assume that to be the case yes, but if so that's been a pretty big oversight on our part with respect to C# projects generally.

Reply:
Ahh, sorry, I misread this comment.

Reply:
We could add an integration test to check this.

Reply:
I think it's not possible to pin multiple versions of dotnet in global.json. So maybe whoever would need this, adds a build script which takes care of the installation.

Reply:
According to the [docs for `global.json`](https://learn.microsoft.com/en-us/dotnet/core/tools/global-json), the version specified there just controls which version of the CLI is used and is independent of the SDK versions that the projects target.